{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gen3 seq bam detected\n"
     ]
    }
   ],
   "source": [
    "import logging\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "import time\n",
    "import argparse\n",
    "import json\n",
    "import pysam\n",
    "import hashlib\n",
    "import subprocess\n",
    "\n",
    "\n",
    "plat={\n",
    "    \n",
    "    'PACBIO':'gen3',\n",
    "    'SEQUEL':'gen3',\n",
    "    'ILLUMINA':'gen2',\n",
    "    'MGISEQ':'gen2'\n",
    "}\n",
    "\n",
    "#input\n",
    "\n",
    "location1=\"/data/yangxiaoxia/bqsr.bam\"\n",
    "location2=\"/home/kechanglin/gen3ccs_new3.ccs.bam\"\n",
    "location=\"/home/kechanglin/data/new_test.bam\" #test input of bam's directory\n",
    "md5_input=\"2cb38082d6d46d425cb7181665e38147\" #md5 of the location file\n",
    "qualimap_loc=\"/home/kechanglin/biosoft/qualimap_v2.2.1/qualimap\"\n",
    "qualimap_out=\"/home/kechanglin/data\"\n",
    "vsf_loc=\"/home/kechanglin/picard.jar\"\n",
    "ccs_out=\"/home/kechanglin/gen3ccs_new3.ccs.bam\"\n",
    "fastq_out=\"/home/kechanglin/data/newfq.fastq\"\n",
    "fastp_out=\"/home/kechanglin/data/fastp_newfq.fq\"\n",
    "\n",
    "\n",
    "def run_cmd(cmd):\n",
    "    pipe = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE,stderr=subprocess.PIPE)\n",
    "    stdout,stderr = pipe.communicate()\n",
    "    pipe.wait()\n",
    "    if pipe.returncode != 0:\n",
    "        raise ValueError(\"Failed to run command :%s, error mesages: %s.\" % (cmd, pipe.stderr.read().decode('utf-8')))\n",
    "    else:\n",
    "        return stdout,stderr\n",
    "    \n",
    "def run_ccs(input_file,output_file,env_name='py27'):\n",
    "    pipe1=subprocess.Popen('source activate '+env_name+' && '+'ccs '+input_file+' '+output_file+' && '+'conda deactivate',shell=True,stdout=subprocess.PIPE,stderr=subprocess.PIPE)\n",
    "    stdout,stderr = pipe1.communicate()\n",
    "    pipe1.wait()\n",
    "    #ccs=run_cmd('source activate '+env_name+' && '+'ccs '+input_file+' '+output_file+' && '+'conda deactivate')\n",
    "    return pipe1.returncode,stdout,stderr\n",
    "\n",
    "\n",
    "class FileMD5(object):\n",
    "    \"\"\"Generate md5.\"\"\"\n",
    "\n",
    "    def __init__(self, filepath):\n",
    "        \"\"\"Init class.\"\"\"\n",
    "        filepath = os.path.abspath(filepath)\n",
    "        if not os.path.isfile(filepath):\n",
    "            raise ValueError(\"Can not find file %s!\" % filepath)\n",
    "        self.filepath = filepath\n",
    "\n",
    "    @property\n",
    "    def md5(self):\n",
    "        \"\"\"Get md5 of file.\"\"\"\n",
    "        hash_md5 = hashlib.md5()\n",
    "        with open(self.filepath, \"rb\") as f:\n",
    "            for chunk in iter(lambda: f.read(4096), b\"\"):\n",
    "                hash_md5.update(chunk)\n",
    "        return hash_md5.hexdigest()\n",
    "\n",
    "    def write_md5(self, outfile):\n",
    "        \"\"\"Write md5 of file to outfile.\"\"\"\n",
    "        with open(outfile, 'wt') as wt:\n",
    "            wt.write('%s  %s' % (self.md5, self.filepath))\n",
    "\n",
    "    def md5check(self, md5):\n",
    "        \"\"\"Check md5 file.\"\"\"\n",
    "        if self.md5 != md5:\n",
    "            return False\n",
    "        else:\n",
    "            return True\n",
    "\n",
    "#maybe useless code\n",
    "def md5sum(md5file):\n",
    "    \"\"\"Md5sum -c file.\"\"\"\n",
    "    md5, filename = ('',) * 2\n",
    "    with open(md5file) as wt:\n",
    "        for line in wt:\n",
    "            if line.startswith(\"MD5\"):\n",
    "                filename, md5 = line.split('=')\n",
    "                filename = filename.replace('MD5(', '').replace(')', '')\n",
    "            else:\n",
    "                md5, filename = line.split('  ')[:2]\n",
    "            md5 = md5.strip()\n",
    "            filename = filename.strip()\n",
    "            filemd5 = FileMD5(filename)\n",
    "            infor = 'succeed' if filemd5.md5check(md5) else 'fail'\n",
    "            print('{} md5 check: {}'.format(filename, infor))\n",
    "\n",
    "def check_md5(filepath, md5):\n",
    "    \"\"\"Check File md5. execution func\"\"\"\n",
    "    filemd5 = FileMD5(filepath)\n",
    "    return filemd5.md5check(md5)\n",
    "        \n",
    "    \n",
    "    \n",
    "    \n",
    "class BAMinput(object):\n",
    "    \"\"\"input bam file and check format\"\"\"\n",
    "    \n",
    "    def __init__(self,directory):\n",
    "        self.directory=directory\n",
    "        if not os.path.isfile(self.directory):\n",
    "            raise IOError(\"no such file %s\" % self.directory)\n",
    "        if not self.directory.endswith('.bam'):\n",
    "            raise IOError(\"unable to manage file other than .bam\")\n",
    "    \n",
    "    def ValidateBAM(self,crash=False):\n",
    "        \n",
    "        \"\"\"format validation and if fatal error met then check md5\"\"\"\n",
    "        self.valinfo=os.popen(\"java -jar \"+vsf_loc+\" ValidateSamFile I=\"+self.directory+\" MODE=SUMMARY\")\n",
    "        for i in self.valinfo:\n",
    "            if not re.match('ERROR',str(i))== None:\n",
    "                print(i)\n",
    "                self.crash=True\n",
    "        if self.crash==True:\n",
    "            self.md5_result=check_md5(self.directory,md5_input)\n",
    "            if self.md5_result:\n",
    "                print('md5 check ok but found fatal error in file format!') #格式有问题，md5无误\n",
    "            else: \n",
    "                raise ValueError('fatal error in file due to failed md5 check!') #格式有问题，md5改变\n",
    "        \n",
    "        \n",
    "    \n",
    "    def gen_identification(self):\n",
    "        self.treatment=False\n",
    "        self.BAMheader=pysam.view(\"-H\",self.directory)\n",
    "        self.platform=str(re.findall(r'PL:\\w+',self.BAMheader))[5:-2] #output platform name\n",
    "        # print(self.platform)\n",
    "        if self.platform == ''or self.platform is None:\n",
    "            print('no equipment info provided, thus unable to tell sequencing technology, it could be a converted file')\n",
    "        elif plat[str.upper(self.platform)]=='gen3':\n",
    "            self.treatment=True #treatment arg is for furture gen3 file handling\n",
    "            print('gen3 seq bam detected')\n",
    "        elif plat[str.upper(self.platform)]=='gen2':\n",
    "            print('gen2 seq bam detected')\n",
    "        else: \n",
    "            print('no recognizable equipment info detected, thus unable to tell sequencing technology, it could be a converted file')\n",
    "        return self.treatment\n",
    "    \n",
    "    \n",
    "    def gen2_qc(self):\n",
    "        self.qc_sub_run=run_cmd(qualimap_loc+' bamqc -bam '+self.directory+' -outdir '+qualimap_out+' -outformat PDF:HTML')\n",
    "        self.run_list=str(self.qc_sub_run[0]).split('\\\\n')\n",
    "        self.show_run= self.run_list[20:42]\n",
    "        return self.show_run\n",
    "        \n",
    "    \n",
    "    def gen3_qc(self):\n",
    "        self.gen3_run=run_ccs(self.directory,ccs_out)\n",
    "        self.gen3_run_log=self.gen3_run[2]\n",
    "        self.fq_log=os.system('samtools fastq '+ccs_out+' > '+fastq_out)\n",
    "        self.fp_log=run_cmd('fastp -i '+fastq_out+' -o '+fastp_out)\n",
    "        self.fastp_report_dir=os.getcwd()\n",
    "        return self.gen3_run_log,self.fq_log,self.fp_log,self.fastp_report_dir\n",
    "\n",
    "\n",
    "        \n",
    "\n",
    "if __name__ == '__main__':\n",
    "#test command    \n",
    "    testinstance=BAMinput(location2) #change this arg to run test\n",
    "    id_treatment=testinstance.gen_identification()\n",
    "    if id_treatment==True:\n",
    "        ccs_result=testinstance.gen3_qc()\n",
    "        # output gen3 result to json\n",
    "    else:\n",
    "        testinstance.ValidateBAM()\n",
    "        gen2_reslut=testinstance.gen2_qc()\n",
    "        # output gen2_result to json\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    #for i in range(len(self.show_run)):\n",
    "         #   print(self.show_run[i])\n",
    "    # check_md5(location,md5_input)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
